### ISSUE TITLE
Compilation error while including the library as submodule

### ISSUE THREAD

**Issue by Priyadharshini-Prabhakaran** [alibaba/yalantinglibs#198]
### Search before asking

- [x] I searched the [issues](https://github.com/alibaba/yalantinglibs/issues) and found no similar issues.

### What happened + What you expected to happen
I am creating a example RPC server application after copy pasting the installed yalantinglibs library into my required directory of my existing SDK.

I am expecting it to compile successfully and produce a rpc_server binary but it causes the following error

``/usr/bin/ld: CMakeFiles/rpc_server.dir/src/rpc_server.cpp.o: in function `coro_rpc::coro_connection::_ZN8coro_rpc15coro_connection10start_implINS_6config17coro_rpc_protocolEN4asio19basic_stream_socketINS4_2ip3tcpENS4_15any_io_executorEEEEEN12async_simple4coro4LazyIvEERNS_8internal6routerIT_EERT0_.actor(coro_rpc::coro_connection::start_impl<coro_rpc::config::coro_rpc_protocol, asio::basic_stream_socket<asio::ip::tcp, asio::any_io_executor> >(coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>&, asio::basic_stream_socket<asio::ip::tcp, asio::any_io_executor>&)::_ZN8coro_rpc15coro_connection10start_implINS_6config17coro_rpc_protocolEN4asio19basic_stream_socketINS4_2ip3tcpENS4_15any_io_executorEEEEEN12async_simple4coro4LazyIvEERNS_8internal6routerIT_EERT0_.frame*)':
/home/sysadmin/my_lib/new_rpc/include/yalantinglibs/coro_rpc/coro_rpc/coro_connection.hpp:181: undefined reference to `coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>::get_handler[abi:cxx11](unsigned int)'
/usr/bin/ld: /home/sysadmin/my_lib/new_rpc/include/yalantinglibs/coro_rpc/coro_rpc/coro_connection.hpp:190: undefined reference to `std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>::route<std::function<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)>*>(unsigned int, std::function<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)>*, std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)'
/usr/bin/ld: /home/sysadmin/my_lib/new_rpc/include/yalantinglibs/coro_rpc/coro_rpc/coro_connection.hpp:184: undefined reference to `coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>::get_coro_handler[abi:cxx11](unsigned int)'
/usr/bin/ld: /home/sysadmin/my_lib/new_rpc/include/yalantinglibs/coro_rpc/coro_rpc/coro_connection.hpp:240: undefined reference to `async_simple::coro::Lazy<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>::route_coro<std::function<async_simple::coro::Lazy<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > (std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)>*>(unsigned int, std::function<async_simple::coro::Lazy<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > (std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)>*, std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)'
collect2: error: ld returned 1 exit status
make[3]: *** [server/CMakeFiles/rpc_server.dir/build.make:84: server/rpc_server] Error 1
make[2]: *** [CMakeFiles/Makefile2:3790: server/CMakeFiles/rpc_server.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:3797: server/CMakeFiles/rpc_server.dir/rule] Error 2
make: *** [Makefile:1542: rpc_server] Error 2``

### Reproduction way
To reproduce use the following step,

1. Clone the latest yalantinglibs library from github
2. Build the application and install the same into /usr/local/include folder using `make install` command
3. Create a new folder inside your existing SDK and copy paste all the installed headers and cmake into a new folder of the SDK
```
mylib
|__CMakeLists.txt
|__new_rpc
    |__include
        |__yalantinglibs
	    |__asio
            |__async_simple
            |__coro_rpc
            |__asio_util
            |__easylog
            |__frozen
            |__iguana
            |__struct_json
	    |__struct_pack
	    |__struct_pb
	    |__util
	    |__asio.hpp
    |__coro_rpcTargets.cmake
    |__easylogTargets.cmake
    |__struct_jsonTargets.cmake
    |__struct_packTargets.cmake
    |__struct_pb.cmake
    |__struct_pbTargets.cmake
    |__yalantinglibsConfig.cmake
    |__yalantinglibsConfigVersion.cmake
|__server
    |__src
        |__rpc_server.cpp
    |__include
        |__service.hpp
|__Other_contents_of_SDK
```    
    
5. Create a sample rpc server application using the APIs from yalantinglibs library 
```
#include "coro_rpc/coro_rpc_server.hpp"
#include "service.h"
using namespace coro_rpc;
void hi() {
 std::cout << "call hi" << std::endl;
}
int main() {
 coro_rpc::coro_rpc_server server(2, 8801);
  //coro_rpc::register_handler<hi>();
  auto ec = server.start();
return ec == std::errc{};
}
```

6. In the main CMakeLists.txt file added the below lines

```
include(${CMAKE_SOURCE_DIR}/new_rpc/yalantinglibsConfig.cmake) include(${CMAKE_SOURCE_DIR}/exc_rpc_new/yalantinglibsConfigVersion.cmake)
add_executable (rpc_server "server/src/rpc_server.cpp" "server/include/service.h"
set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fcoroutines")
target_compile_definitions(rpc_server PRIVATE ASIO_HAS_STD_INVOKE_RESULT)
target_compile_definitions(rpc_server PRIVATE ASYNC_SIMPLE_HAS_NOT_AIO)
target_link_libraries(rpc_server PRIVATE yalantinglibs::struct_json yalantinglibs::coro_rpc)
```
7. Now build the application and you will get the above error

### Are you willing to submit a PR?

- [x] Yes I am willing to submit a PR!

**Comment by PikachuHyA** [comment]
sorry for late reply

1. why add this code to CMakeLists.txt?

```
include(${CMAKE_SOURCE_DIR}/new_rpc/yalantinglibsConfig.cmake) include(${CMAKE_SOURCE_DIR}/exc_rpc_new/yalantinglibsConfigVersion.cmake)
```

2. doc: use as git submodule

see the demo: https://alibaba.github.io/yalantinglibs/guide/how-to-use-as-git-submodule.html

2. use by find_package

after `make install`, I recomand  use CMake `find_package`

see https://github.com/alibaba/yalantinglibs/pull/177

```
set(CMAKE_CXX_STANDARD 20)
find_package(yalantinglibs  REQUIRED)
add_executable(your_target main.cpp)
target_link_libraries(your_target PRIVATE yalantinglibs::coro_rpc)
```

I will add the doc later.

**Comment by PikachuHyA** [comment]
use by cmake find_package

https://alibaba.github.io/yalantinglibs/guide/how-to-use-by-cmake-find_package.html

**Comment by Priyadharshini-Prabhakaran** [comment]
Thank you for the response.

I tried creating a separate project and followed the below steps,

1. Cloned yalantinglibs library, build and installed the same
2. Created the following folder structure:
```
test_project
  |__rpc_server.cpp
  |__service.h
  |__CMakeLists.txt
```
3. File contents:

rcp_server.cpp
```
#include "coro_rpc/coro_rpc_server.hpp"
#include "service.h"

using namespace coro_rpc;
void hi() {
  std::cout << "call hi" << std::endl;
}
int main() {
  coro_rpc::coro_rpc_server server(2, 8801);
  //coro_rpc::register_handler<hi>();
  auto ec = server.start();
  return ec == std::errc{};
}
```

service.h
```
#pragma once
void hi();
```

CMakeLists.txt
```
cmake_minimum_required(VERSION 3.15)

project(awesome-solution LANGUAGES CXX)
set(CMAKE_CXX_STANDARD 20)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
# it's very important to set the compiler flags
# if you use gcc/g++
if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fcoroutines")
    if (CMAKE_CXX_COMPILER_VERSION MATCHES "12.*")
        set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-maybe-uninitialized")
    endif()
endif()
set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
find_package(Threads REQUIRED)
find_package(yalantinglibs REQUIRED)
# add your server
add_executable(server_app rpc_server.cpp)
target_link_libraries(server_app PRIVATE yalantinglibs::coro_rpc)
target_compile_definitions(server_app PRIVATE ASIO_HAS_STD_INVOKE_RESULT)
target_compile_definitions(server_app PRIVATE ASYNC_SIMPLE_HAS_NOT_AIO)
```
4. When I build the project I am still getting the below error:

```
/usr/bin/ld: CMakeFiles/server_app.dir/rpc_server.cpp.o: in function `coro_rpc::coro_connection::_ZN8coro_rpc15coro_connection10start_implINS_6config17coro_rpc_protocolEN4asio19basic_stream_socketINS4_2ip3tcpENS4_15any_io_executorEEEEEN12async_simple4coro4LazyIvEERNS_8internal6routerIT_EERT0_.actor(coro_rpc::coro_connection::start_impl<coro_rpc::config::coro_rpc_protocol, asio::basic_stream_socket<asio::ip::tcp, asio::any_io_executor> >(coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>&, asio::basic_stream_socket<asio::ip::tcp, asio::any_io_executor>&)::_ZN8coro_rpc15coro_connection10start_implINS_6config17coro_rpc_protocolEN4asio19basic_stream_socketINS4_2ip3tcpENS4_15any_io_executorEEEEEN12async_simple4coro4LazyIvEERNS_8internal6routerIT_EERT0_.frame*)':
rpc_server.cpp:(.text+0x4916): undefined reference to `coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>::get_handler[abi:cxx11](unsigned int)'
/usr/bin/ld: rpc_server.cpp:(.text+0x4980): undefined reference to `coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>::get_coro_handler[abi:cxx11](unsigned int)'
/usr/bin/ld: rpc_server.cpp:(.text+0x4a81): undefined reference to `async_simple::coro::Lazy<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>::route_coro<std::function<async_simple::coro::Lazy<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > (std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)>*>(unsigned int, std::function<async_simple::coro::Lazy<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > (std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)>*, std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)'
/usr/bin/ld: rpc_server.cpp:(.text+0x4cc9): undefined reference to `std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > coro_rpc::internal::router<coro_rpc::config::coro_rpc_protocol>::route<std::function<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)>*>(unsigned int, std::function<std::pair<std::errc, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > (std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)>*, std::basic_string_view<char, std::char_traits<char> >, std::shared_ptr<coro_rpc::coro_connection>, std::variant<coro_rpc::config::struct_pack_protocol>)'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/server_app.dir/build.make:84: server_app] Error 1
make[1]: *** [CMakeFiles/Makefile2:76: CMakeFiles/server_app.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
```

Kindly help me out with the same.

**Comment by poor-circle** [comment]
> 

Could you pull the code from main then try again? It seems fixed by #202.
